install-template.sh

```diff
@@ -145,11 +145,16 @@ append_to_file() {
 
 make_dir_recursive() {
     local _dir="$1"
-    local _line="$ umask 022 && mkdir -p \"$_dir\""
-    umask 022 && mkdir -p "$_dir"
-    local _retval=$?
-    log_line "$_line"
-    return $_retval
+    # Skip if the last invocation of make_dir_recursive had the same argument
+    if ! [ "$_dir" = "${_make_dir_recursive_cached_key:-}" ]
+    then
+        _make_dir_recursive_cached_key="$_dir"
```

Comment:
This seems like kind of an odd optimization, it only works if the very last call was the same directory. Have you compared the times with and without it?

Reply:
Yes, see commit message https://github.com/rust-lang/rust-installer/pull/106/commits/53d557be7e6e30f4431acf31cfb86d2b01317d55 : input lists are sorted by path, the directories usually contain more than one file and as such this optimization is hit frequently.
I did compare times and the difference was significant enough to warrant the added complexity. However, I did not record any numbers.

Reply:
This also significantly reduced the log file size.

Reply:
Went AFK for a bit to let this run:
```
# pwd              
/tmp/rust-1.51.0-x86_64-unknown-linux-gnu
# stat -fc%T /tmp                      
tmpfs
# dst=/tmp/destdir                                                                                                                                                                                                
# for c in 53d557b 3d7ed69 5254dbf ; do curl -sLO "https://raw.githubusercontent.com/mbargull/rust-installer/${c}/install-template.sh" ; rm -rf "${dst}" ; printf %s\\n "${c}" ; time bash install-template.sh --destdir="${dst}" >/dev/null ; find "${dst}" -name \*.log -exec stat -c'%n %s' {} \; ; done
53d557b
install: WARNING: failed to run ldconfig. this may happen when not installing as root. run with --verbose to see the error
real 44.19
user 35.54
sys 11.26
/tmp/destdir/usr/local/lib/%%TEMPLATE_REL_MANIFEST_DIR%%/install.log 11361126
3d7ed69
install: WARNING: failed to run ldconfig. this may happen when not installing as root. run with --verbose to see the error
real 105.68
user 82.46
sys 30.43
/tmp/destdir/usr/local/lib/%%TEMPLATE_REL_MANIFEST_DIR%%/install.log 12935429
5254dbf
install: WARNING: failed to run ldconfig. this may happen when not installing as root. run with --verbose to see the error
real 688.47
user 570.21
sys 207.40
/tmp/destdir/usr/local/lib/%%TEMPLATE_REL_MANIFEST_DIR%%/install.log 12935429
```
So for me that commit cuts about 60 % of the time on a `tmpfs`.
(`install.log` size difference is not that big in this case due to short prefix/destdir. At https://github.com/conda-forge/rust-feedstock we have longer (\~ 255 characters) install prefixes, resulting in a much larger log file.)

Reply:
Hmm, I wonder if you could get almost as much of a speedup by removing the logging. It sounds like it takes a while and isn't super useful.

Reply:
No, the logging is just uses the `echo` builtin which should be able to offer high throughput.
One thing to try is whether we can just use a `test -d DIR` for `make_dir_recursive` instead of the memorized logic. (For the abs path function I don't see a similarly performing alternative.)

Reply:
Yep, `test -d` gives the same run time. I've added two cleanup commits.

Reply:
(Never mind the 2nd commit which I force-pushed away. I just forgot why (abs path from file vs dir) I split those functions.)